targets/csharp/source/source/PlayFabGameServerInternal.cs

```diff
@@ -0,0 +1,310 @@
+            TransitionToActiveEvent = new ManualResetEvent(false);
+        }
+
+        public Task StartAsync(bool shouldWriteDebugLogs = false)
+        {
+            // If we already initialized everything, no need to do it again
+            if (_heartbeatTask != null)
+            {
+                return Task.CompletedTask;
+            }
+
+            _shouldWriteDebugLogs = shouldWriteDebugLogs;
+
+            _configuration = _configuration ?? GetConfiguration();
+            ConfigMap = ConfigMap ?? CreateConfigMap(_configuration);
+            Logger = LoggerFactory.CreateInstance(ConfigMap[PlayFabGameServerSDK.LogFolderKey]);
+            
+            if (_configuration.ShouldLog())
+            {
+                Logger.Start();
+            }
+
+            string gsmsBaseUrl = ConfigMap[PlayFabGameServerSDK.HeartbeatEndpointKey];
+            string instanceId = ConfigMap[PlayFabGameServerSDK.ServerIdKey];
+
+            Logger.Log($"VM Agent Endpoint: {gsmsBaseUrl}");
+            Logger.Log($"Instance Id: {instanceId}");
+
+            _webClient = GameServerHttpClientFactory.CreateInstance($"http://{gsmsBaseUrl}/v1/sessionHosts/{instanceId}");
+
+            _heartbeatRunning = true;
+            _signalHeartbeatEvent.Reset();
+            TransitionToActiveEvent.Reset();
+
+            _heartbeatTask = Task.Run(HeartbeatAsync);
+            return Task.CompletedTask;
+        }
+
+        private GameServerConfiguration GetConfiguration()
+        {
+            string fileName = !string.IsNullOrWhiteSpace(_overrideConfigFileName) ? _overrideConfigFileName
+                                : Environment.GetEnvironmentVariable(PlayFabGameServerSDK.GsdkConfigFileEnvVarKey);
+
+            GameServerConfiguration localConfig;
+
+            if (!string.IsNullOrWhiteSpace(fileName) && File.Exists(fileName))
+            {
+                var jsonWrapper = PlayFab.PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer);
+
+                localConfig = new JsonFileConfiguration(jsonWrapper, fileName);
+            }
+            else
+            {
+                localConfig = new EnvironmentVariableConfiguration();
+            }
+
+            return localConfig;
+        }
+
+        private IDictionary<string, string> CreateConfigMap(GameServerConfiguration localConfig)
+        {
+            var finalConfig = new Dictionary<string, string>();
+
+            foreach (KeyValuePair<string, string> certEntry in localConfig.GameCertificates)
+            {
+                finalConfig[certEntry.Key] = certEntry.Value;
+            }
+
+            foreach (KeyValuePair<string, string> metadata in localConfig.BuildMetadata)
+            {
+                finalConfig[metadata.Key] = metadata.Value;
+            }
+
+            foreach (KeyValuePair<string, string> port in localConfig.GamePorts)
+            {
+                finalConfig[port.Key] = port.Value;
+            }
+
```

Comment:
I remember there has been some discussion in the past (at least around metadata) about whether the keys are case sensitive or not. Is that something that we should be careful about here as well?

Reply:
the idea here is anyone using the GSDK should use the pre-defined constants (so they don't have to worry about mismatch in the key casing) but I'm considering just using a configuration object here

---
In reply to: [256217055](https://github.com/PlayFab/SDKGenerator/pull/398#discussion_r256217055) [](ancestors = 256217055)